File: ABC_toy/R_version/toy.R
require(EasyABC)
require(ggplot2)
# Toy programm


# ======================================================================
#   COMPUTE PREDICTED VALUES: BLACK BOX
# ======================================================================
#     the numbers in the following equations are "hidden"
toy_model <- function(par){
  val1 <- 1e-1*par[1]^4 +
          2e-1*par[2]^3.0 +
          7e-1*par[3]^2 +
          4e-1*par[4];
  val2 <- 3e-2*par[1]^4.0 +
          5e-2*par[2]^3.0 +
          9e-2*par[3]^2.0 + 
          1e-2*par[4];
  val3 <- 7e-3*par[1]^4.0 +
          1e-3*par[2]^3.0 +
          4e-3*par[3]^2 +
          2e-3*par[4];
  val4 <- 2e-4*par[1]^4 +
          5e-4*par[2]^3 +
          8e-4*par[3]^2 +
          8e-4*par[4]
  val5 <- 9e-5*par[1]^4 +
          6e-5*par[2]^3 +
          3e-5*par[3]^2 +
          1e-5*par[4]
  

return(c(val1,val2,val3,val4,val5))

}



toy_model <- cmpfun(toy_model)
# Observed data
x1 = 34
x2 = 12
x3 = 456
x4 = 78

obs_data <- toy_model(c(x1,x2,x3,x4))
# Summary, just the data - no compression
sum_stat_obs <- obs_data

toy_prior <- list(c("lognormal",log(2.5),log(2.5)),c("lognormal",log(2.5),log(2.5)),
                  c("lognormal",log(2.5),log(2.5)),c("lognormal",log(2.5),log(2.5)))


# Fraction of acceptable data, measured via Euclidian distance between simulation and obs_data
tolerance=c(1,0.75,0.5,0.25,0.15)




n=1e5
tol = 0.1
ABC_sim<-ABC_rejection(model=toy_model, prior=toy_prior, nb_simul=n, summary_stat_target=sum_stat_obs,tol=tol,
use_seed=TRUE)

hist(ABC_sim$param[,4])

pdat <- as.data.frame(ABC_sim$param)

ggplot(data = pdat,aes(x = V1,y = V3)) +
  geom_point()


# Run Sequencial Approximate Bayesian Computation
ABC_Beaumont <- ABC_sequential(method="Beaumont", model=toy_model,
                             prior=toy_prior, nb_simul=1e4, summary_stat_target=sum_stat_obs,
                             tolerance_tab=tolerance)

# Plot histogram
hist(ABC_Beaumont$param[,1])

# Plot density 2D
pdat <- as.data.frame(ABC_Beaumont$param)

ggplot(data = pdat,aes(x = V1,y = V2)) +
  geom_density_2d()

File: GA_toy/toy.f
      subroutine toy(n,par,y)
c
c     this taked the place of nucleo.f. It reads parameters from an
c     input file and outputs simulated values from these parameters   
c
c     with the following 4 parameters:
c
c     34         ! par_1
c     12         ! par_2
c     456        ! par_3
c     78         ! par_4
c
c     we obtain the 5 simulated values:
c
c     2.80E+05   ! val1
c     5.89E+04   ! val2
c     1.02E+04   ! val3
c     4.35E+02   ! val4
c     1.27E+02   ! val5
c
c======================================================================
c   VARIABLES
      integer n,m
      real*8 par(n)
      real*8 y(m)
      
c   COMPUTE PREDICTED VALUES: BLACK BOX
c======================================================================
c     the numbers in the following equations are "hidden"
      y(1)=1.d-1*par(1)**4.0d0+
     &     2.d-1*par(2)**3.0d0+
     &     7.d-1*par(3)**2.0d0+
     &     4.d-1*par(4)
      y(2)=3.d-2*par(1)**4.0d0+
     &     5.d-2*par(2)**3.0d0+
     &     9.d-2*par(3)**2.0d0+
     &     1.d-2*par(4)
      y(3)=7.d-3*par(1)**4.0d0+
     &     1.d-3*par(2)**3.0d0+
     &     4.d-3*par(3)**2.0d0+
     &     2.d-3*par(4)
      y(4)=2.d-4*par(1)**4.0d0+
     &     5.d-4*par(2)**3.0d0+
     &     8.d-4*par(3)**2.0d0+
     &     8.d-4*par(4)
      y(5)=9.d-5*par(1)**4.0d0+
     &     6.d-5*par(2)**3.0d0+
     &     3.d-5*par(3)**2.0d0+
     &     1.d-5*par(4)
      
      return
      end

